Fix ambiguous data issues

oxidase opened the pull request:
Added disambiguations for edge weights and names by taking minimal edge weight and minimal name id for edges with equal source and target nodes.

Solves  #2617

TheMarex commented on src/extractor/extraction_containers.cpp:
```diff
@@ -453,11 +453,13 @@ void ExtractionContainers::PrepareEdges(lua_State *segment_state)
                 all_edges_list[i].result.weight < min_forward_weight)
             {
                 min_forward_idx = i;
+                min_forward_weight = all_edges_list[i].result.weight;
```
This would mean the de-duplication code has been broken for a while?

oxidase replied:
Yes, always the last edge was used. Because the order of edges was (source, target),  edge weights were "randomized" by scheduler.

MoKob replied:
this value is never read? And the IDX is used to choose only the minimal edge. All others are invalidated. Does this even affect anything?

oxidase replied:
Yes, without the change min_forward_weight is always std::numeric_limits<int>::max()
and "all_edges_list[i].result.weight < min_forward_weight" is always true

MoKob replied:
Ah yes. You are right. Loops and no coffee don't seem to mix.

MoKob replied:
Can we add a test-case with multiple parallel speed-edges and empty name and check for speed?

oxidase replied:
Ok, I have added empty names to the test

TheMarex commented on include/extractor/internal_extractor_edge.hpp:
```diff
@@ -142,14 +142,16 @@ struct InternalExtractorEdge
-                (lhs.result.target < rhs.result.target));
+                (lhs.result.target < rhs.result.target ||
+                 (lhs.result.target == rhs.result.target &&
+                  lhs.result.name_id < rhs.result.name_id)));
```
Usually having a name is better than not having a name so inverting this is probably more what we want.

oxidase replied:
Ok, i will revert the order.

TheMarex commented:
Good catch. This must have been a very old bug, at least going back a year.

systemed commented:
Very interesting. I've occasionally found issues similar to #610 where OSRM appears to pick the less favourable way between two nodes - for example, at http://cycle.travel/map?from=51.4691,-0.2196&to=51.4707,-0.2216, where the middle section is seemingly following a highway=pedestrian area even though the road has a better weight. Hopefully this fixes it.

oxidase commented:
@systemed looks like it can be solved by PR.

emiltin commented:
Would be really nice if this could solve #610 - which is over 3 years old :-)

MoKob commented on include/extractor/internal_extractor_edge.hpp:
```diff
@@ -142,14 +142,16 @@ struct InternalExtractorEdge
-                (lhs.result.target < rhs.result.target));
+                (lhs.result.target < rhs.result.target ||
+                 (lhs.result.target == rhs.result.target &&
+                  lhs.result.name_id > rhs.result.name_id)));
```
the parsing happens asynchronously, doesn't it? So the edges arrive in a different order which should also result in name_ids being assigned randomly? Or am I missing something? NameID should not be possible to use here?

oxidase replied:
Yes, it fixes only the original issue with an empty name and non-empty one.

MoKob replied:
Well... fixes. It randomly chooses any of the non-empty names over an empty name. In that case, your test-case is still wrong, since it offers multiple possible names (50:50 chance)

MoKob commented on features/testbot/basic.feature:
```diff
@@ -251,3 +251,34 @@ Feature: Basic Routing
+        When I route I should get
+            | from | to | route |
+            | a    | c  | 2,3,3 |
+            | c    | a  | 3,2,2 |
```
Could fail randomly, due to parallel parsing. `ab` are all identical, so they are differentiated on by the name_id which is random due to https://github.com/Project-OSRM/osrm-backend/blob/0e4061f85816b5499896022eb151a1402ae3aad0/src/extractor/extractor.cpp#L212 and its interaction with https://github.com/Project-OSRM/osrm-backend/blob/0e4061f85816b5499896022eb151a1402ae3aad0/src/extractor/extractor.cpp#L160

oxidase replied:
Frankly, i have no idea what to use for stable ordering of names. Would it make sense to take a lexicographically  smallest non-empty string?

MoKob replied:
I think that might be an option. Otherwise we end up with just another undefined behaviour.

oxidase replied:
Ok, i will update PR with as suggested

oxidase commented:
Added a lexicographical compare for names to disambiguate edges. Empty strings are sorted as highest values. The price is additional memory used for "STXXLNameOffsets name_offsets".
I also changed data type to "unsigned char" because  the utf8 comparison with  char code units will sort strings incorrectly   in order Greek, Cyrillic, Latin.

danpat commented on src/extractor/extraction_containers.cpp:
```diff
@@ -53,6 +112,12 @@ ExtractionContainers::ExtractionContainers()
+    name_offsets.push_back(0);
+    name_offsets.push_back(0);
+    name_offsets.push_back(0);
+
```
Looks like there's one extra here?

oxidase replied:
No, the last offset is always the total length of the character data array and at the moment it is zero. Offsets are added after pushing data into the data  array.

MoKob commented on src/extractor/extraction_containers.cpp:
```diff
@@ -53,6 +112,12 @@ ExtractionContainers::ExtractionContainers()
     name_lengths.push_back(0);
     name_lengths.push_back(0);
 
+    // Insert three zero offsets for empty strings and the total length
```
this comment needs to be adjusted, if we push four zeros. It also requires an explanation of what the different indices are supposed to be.

oxidase commented on src/extractor/extraction_containers.cpp:
```diff
@@ -35,6 +37,63 @@ struct OSMNodeIDSTXXLLess
+            return true;
+
+        BOOST_ASSERT(!name_offsets.empty() && name_offsets.back() == name_data.size());
+        const oe::ExtractionContainers::STXXLNameCharData::const_iterator data = name_data.begin();
```
@MoKob I made the original comparison based on name ids due to performance and efforts, but it is not correct anyway, because if there more than one non-empty names for one road the scheduler will randomize names and it will be a problem for the graph compressor and guidance to decide that it is the same road.

The statistic must be checked for such cases. For Jakarta case it is below 0.1% when it gets to names comparison, so performance impact is not so dramatic.

TheMarex replied:
We can run some statistics on this. My feeling is that branch is almost never hit. This will only be used if there are multi-edges in OSM which is quite rare. Having consistent tie breaking is definitely desirable.

MoKob replied:
Well. We can always risk it... Its just a chance of a potential performance bottle-neck. Maybe add some commentary around this to make clear that this could impact extract performance.

The problem why I insisted on having clear distinction is the parallel-edges tests. Because they could fail randomly without this destinction. So if we have tests that have this randomness-possibility, we need to get rid of randomness.

oxidase commented:
I have updated PR to avoid additional memory usage. Bad place now is a conversion from offsets to lengths for RangeTable in WriteCharData. Also lexicographical order is not  good for resolving street names ambiguity, and can result in some extremely rare strange street name changes.
For example, if edges have ambiguous names (A,C), (B,C), (D,C), (E,C) than names will be A,B,C,C.

I have also checked frequency of string comparisons: for oberbayern only 4 for 1466927 edges
```
1: Fischer-von-Erlach-Straße Lohensteinstraße
2: Lohensteinstraße Fischer-von-Erlach-Straße
3: Lohensteinstraße Fischer-von-Erlach-Straße
4: Maximilianstraße Römerkesselstraße (St 2055)
```

For Germany 600 comparisons for 24882388 edges
[strings.txt](https://github.com/Project-OSRM/osrm-backend/files/355503/strings.txt)

I don't understand at the moment why so many comparisons  for some street names like "St. Roman", but for me it looks like data problems.

TheMarex commented:
> upd: Added avoiding special node ids, so only 35 string comparisons left for 24882388 edges in Germany

That looks good to me. I don't think this will be noticable.

W.r.t the quality of the tie breaking: While a consitent naming for a street is desirable, I don't think these edge case warrants a more advanced technique.
